File: demo/src/main/java/com/devbrackets/android/exomediademo/service/MediaImageProvider.java
package com.devbrackets.android.exomediademo.service;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;
import com.devbrackets.android.exomediademo.R;
import com.devbrackets.android.exomediademo.data.MediaItem;
import com.devbrackets.android.playlistcore.components.image.ImageProvider;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


public class MediaImageProvider implements ImageProvider<MediaItem> {
    interface OnImageUpdatedListener {
        void onImageUpdated();
    }

    @NotNull
    private RequestManager glide;
    @NonNull
    private OnImageUpdatedListener listener;

    @NonNull
    private NotificationImageTarget notificationImageTarget = new NotificationImageTarget();
    @NonNull
    private RemoteViewImageTarget remoteViewImageTarget = new RemoteViewImageTarget();

    @NonNull
    private Bitmap defaultNotificationImage;

    @Nullable
    private Bitmap notificationImage;
    @Nullable
    private Bitmap artworkImage;

    public MediaImageProvider(@NonNull Context context, @NonNull OnImageUpdatedListener listener) {
        glide = Glide.with(context.getApplicationContext());
        this.listener = listener;

        defaultNotificationImage = BitmapFactory.decodeResource(context.getResources(), R.mipmap.ic_launcher);
    }

    @Override
    public int getNotificationIconRes() {
        return R.mipmap.ic_launcher;
    }

    @Override
    public int getRemoteViewIconRes() {
        return R.mipmap.ic_launcher;
    }

    @Nullable
    @Override
    public Bitmap getLargeNotificationImage() {
        return notificationImage != null ? notificationImage : defaultNotificationImage;
    }

    @Nullable
    @Override
    public Bitmap getRemoteViewArtwork() {
        return artworkImage;
    }

    @Override
    public void updateImages(@NotNull MediaItem playlistItem) {
        glide.asBitmap().load(playlistItem.getThumbnailUrl()).into(notificationImageTarget);
        glide.asBitmap().load(playlistItem.getArtworkUrl()).into(remoteViewImageTarget);
    }

    /**
     * A class used to listen to the loading of the large notification images and perform
     * the correct functionality to update the notification once it is loaded.
     * <p>
     * <b>NOTE:</b> This is a Glide Image loader class
     */
    private class NotificationImageTarget extends SimpleTarget<Bitmap> {
        @Override
        public void onResourceReady(@NonNull Bitmap resource, @Nullable Transition<? super Bitmap> transition) {
            notificationImage = resource;
            listener.onImageUpdated();
        }
    }

    /**
     * A class used to listen to the loading of the large lock screen images and perform
     * the correct functionality to update the artwork once it is loaded.
     * <p>
     * <b>NOTE:</b> This is a Glide Image loader class
     */
    private class RemoteViewImageTarget extends SimpleTarget<Bitmap> {
        @Override
        public void onResourceReady(@NonNull Bitmap resource, @Nullable Transition<? super Bitmap> transition) {
            artworkImage = resource;
            listener.onImageUpdated();
        }
    }
}

File: demo/src/main/kotlin/com/devbrackets/android/exomediademo/playlist/service/MediaImageProvider.kt
package com.devbrackets.android.exomediademo.playlist.service

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestManager
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.transition.Transition
import com.devbrackets.android.exomediademo.R
import com.devbrackets.android.exomediademo.data.MediaItem
import com.devbrackets.android.playlistcore.components.image.ImageProvider

class MediaImageProvider(context: Context, private val onImageUpdated: () -> Unit) : ImageProvider<MediaItem> {
  private val glide: RequestManager by lazy { Glide.with(context.applicationContext) }

  private val notificationImageTarget = NotificationImageTarget()
  private val remoteViewImageTarget = RemoteViewImageTarget()

  private val defaultNotificationImage by lazy {
    BitmapFactory.decodeResource(context.resources, R.drawable.ic_media_notification)
  }

  private var notificationImage: Bitmap? = null
  override var remoteViewArtwork: Bitmap? = null
    private set

  override val notificationIconRes: Int
    get() = R.drawable.ic_media_notification

  override val remoteViewIconRes: Int
    get() = R.drawable.ic_media_notification

  override val largeNotificationImage: Bitmap
    get() = notificationImage ?: defaultNotificationImage

  override fun updateImages(playlistItem: MediaItem) {
    glide.asBitmap().load(playlistItem.thumbnailUrl).into(notificationImageTarget)
    glide.asBitmap().load(playlistItem.artworkUrl).into(remoteViewImageTarget)
  }

  /**
   * A class used to listen to the loading of the large notification images and perform
   * the correct functionality to update the notification once it is loaded.
   *
   * **NOTE:** This is a Glide Image loader class
   */
  private inner class NotificationImageTarget : SimpleTarget<Bitmap>() {
    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
      notificationImage = resource
      onImageUpdated()
    }
  }

  /**
   * A class used to listen to the loading of the large lock screen images and perform
   * the correct functionality to update the artwork once it is loaded.
   *
   * **NOTE:** This is a Glide Image loader class
   */
  private inner class RemoteViewImageTarget : SimpleTarget<Bitmap>() {
    override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
      remoteViewArtwork = resource
      onImageUpdated()
    }
  }
}
